h2/src/main/org/h2/mvstore/tx/Transaction.java

```diff
@@ -120,17 +199,19 @@ public void markStatementEnd() {
      * @param key the key
      * @param oldValue the old value
      */
-    void log(int mapId, Object key, Object oldValue) {
+    void log(int mapId, Object key, VersionedValue oldValue) {
+        long currentState = statusAndLogId.getAndIncrement();
```

Comment:
What if overflow happens here? Status will be damaged too?

Reply:
Hypothetically some extra zero bit between these fields may be used as a guard to detect the overflow.

I don't think that H2 is really capable to handle such large amount of changes, however. But logic should either look sane or be documented somehow to avoid such questions in the future.

Reply:
Sure, but I would consider us lucky if H2 will ever successfully commit any transaction with 2**40-1 changes :smile:

Reply:
Of course you can detect overflow and ... do something. But by the same token, why we are not checking result of EVERY arithmetic or de-reference operation? And the usual answer to all possible questions of this kind is - because we are trying to strike a balance between acceptable performance and negligible risk of failure.

Reply:
I think that should be enough to place here some comment about that number of changes needed to force the overflow is too large to be reachable in practice.

Reply:
We are not shrinking range of possible values here, it was and stays 40 bits, and curious reader can find out-pf-range checks in TransactionStore.getOperationId(), which IMHO serve mostly as comments. We can put similar checks here, just to avoid questions, in the future.